trust-quorum/src/connection_manager.rs

```diff
@@ -0,0 +1,717 @@
+        // connection manager won't work properly without doing this because it
+        // will never trigger connections since its own address will always sort
+        // lower than other addresses if only the ports differ.
+        let listen_port = server.listen_addr().unwrap().port();
+
+        if listen_port != listen_addr.port() {
+            listen_addr.set_port(listen_port);
+        }
+
+        info!(
+            log,
+            "Started listening";
+            "local_addr" => %listen_addr
+        );
+
+        ConnMgr {
+            log,
+            main_tx,
+            config,
+            server,
+            listen_addr,
+            next_task_id: TaskId::new(0),
+            join_handles: Default::default(),
+            bootstrap_addrs: BTreeSet::new(),
+            connecting: BTreeMap::new(),
+            accepting: BTreeMap::new(),
+            established: BTreeMap::new(),
+        }
+    }
+
+    pub async fn shutdown(&mut self) {
+        // Shutdown all connection processing tasks
+        for (_, handle) in &self.accepting {
+            let _ = handle.tx.send(MainToConnMsg::Close).await;
+        }
+        for (_, handle) in &self.connecting {
+            let _ = handle.tx.send(MainToConnMsg::Close).await;
+        }
+        for (_, handle) in &self.established {
+            let _ = handle.tx.send(MainToConnMsg::Close).await;
+        }
+    }
+
+    pub fn status(&self) -> ConnMgrStatus {
+        let connections = self
+            .connecting
+            .iter()
+            .map(|(addr, task_handle)| ConnInfo {
+                state: ConnState::Connecting,
+                addr: *addr,
+                task_id: task_handle.task_id,
+            })
+            .chain(self.accepting.iter().map(|(addr, task_handle)| ConnInfo {
+                state: ConnState::Accepting,
+                addr: *addr,
+                task_id: task_handle.task_id,
+            }))
+            .chain(self.established.iter().map(
+                |(baseboard_id, task_handle)| ConnInfo {
+                    state: ConnState::Established(baseboard_id.clone()),
+                    addr: task_handle.addr(),
+                    task_id: task_handle.task_id,
+                },
+            ))
+            .collect();
+
+        ConnMgrStatus {
+            bootstrap_addrs: self.bootstrap_addrs.clone(),
+            connections,
+            num_task_join_handles: self.join_handles.len() as u64,
+            next_task_id: self.next_task_id,
+        }
+    }
+
+    pub fn listen_addr(&self) -> SocketAddrV6 {
+        self.listen_addr
+    }
+
+    /// Perform any polling related operations that the connection
+    /// manager must perform concurrently.
+    pub async fn step(
+        &mut self,
+        corpus: Vec<Utf8PathBuf>,
+    ) -> Result<(), AcceptError> {
+        let mut interval = interval(RECONNECT_TIME);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
```

Comment:
nitpick, take it or leave it: it seems like the `interval` could be part of `Self` and reused every time we step the thing, so we don't have to create and configure a new one every time?